src/main/java/com/google/cloud/dataproc/templates/dataplex/DataplexGCStoBQ.java

```diff
@@ -0,0 +1,426 @@
+   *
+   * @param args line arguments
+   * @return parsed arguments
+   */
+  public static CommandLine parseArguments(String... args) {
+    Options options = new Options();
+    Option customSQLFileOption =
+        new Option(CUSTOM_SQL_GCS_PATH_OPTION, "GCS path of file containing custom sql");
+    customSQLFileOption.setRequired(false);
+    customSQLFileOption.setArgs(1);
+    options.addOption(customSQLFileOption);
+
+    Option entityListOption = new Option(ENTITY_LIST_OPTION, "Dataplex GCS table resource name");
+    entityListOption.setRequired(false);
+    entityListOption.setArgs(2);
+    options.addOption(entityListOption);
+
+    Option assetOption = new Option(ASSET_LIST_OPTION, "Dataplex asset name");
+    assetOption.setRequired(false);
+    assetOption.setArgs(3);
+    options.addOption(assetOption);
+
+    Option partitionField = new Option(PARTITION_FIELD_OPTION, "BigQuery partitionField");
+    partitionField.setRequired(false);
+    partitionField.setArgs(4);
+    options.addOption(partitionField);
+
+    Option partitionType = new Option(PARTITION_TYPE_OPTION, "BigQuery partitionType");
+    partitionType.setRequired(false);
+    partitionType.setArgs(5);
+    options.addOption(partitionType);
+
+    CommandLineParser parser = new BasicParser();
+    try {
+      return parser.parse(options, args);
+    } catch (ParseException e) {
+      throw new IllegalArgumentException(e);
+    }
+  }
+
+  /**
+   * Parse partitionsListWithLocationAndKeys into a Dataset partition path and key values for each
+   * partition in Dataplex Entity
+   *
+   * @param partitionsListWithLocationAndKeys list with partition path and key values for each
+   *     partition in Dataplex Entity
+   * @param partitionKeysList list of partition keys for Dataplex Entity
+   * @return a dataset with partition path and key values for each partition in Dataplex Entity
+   */
+  public Dataset<Row> getAllPartitionsDf(
+      List<String> partitionsListWithLocationAndKeys, List<String> partitionKeysList) {
+    Dataset allPartitionsDf =
+        sqlContext.createDataset(partitionsListWithLocationAndKeys, Encoders.STRING()).toDF();
+    allPartitionsDf =
+        allPartitionsDf.selectExpr(SPARK_SQL_SELECT_STAR, SPARK_SQL_SPLIT_VALUE_AND_GET_LOCATION);
+    for (int i = 0; i < partitionKeysList.size(); i += 1) {
+      allPartitionsDf =
+          allPartitionsDf.selectExpr(
+              SPARK_SQL_SELECT_STAR,
+              String.format(SPARK_SQL_SPLIT_VALUE_AND_GET_KEY, i + 1, partitionKeysList.get(i)));
+    }
+    allPartitionsDf = allPartitionsDf.drop(SPARK_SQL_VALUE_FIELD);
+    return allPartitionsDf;
+  }
+
+  /**
+   * Query BQ target all distinct value of partition keys
+   *
+   * @param partitionKeysList list of partition keys for Dataplex Entity
+   * @return a dataset with all distinct value of partition keys in BQ target table
+   */
+  private Dataset<Row> getBQTargetAvailablePartitionsDf(List<String> partitionKeysList) {
```

Comment:
Out of curiousity, can't Bigquery APIs be here to fetch information?

Reply:
That's a good point. We did consider that option but it requires Dataplex and BQ partitioning scheme to be identical. This will most likely not be the case in most scenarios, since BigQuery only support partitioning by time-unit columns, ingestion time, and int. A very common case is one where a partition key in Dataplex is of type string, in this case the partitioning scheme of Dataplex and BigQuery could not made to match.